### PR TITLE
Bugfix: resource management

### PR DESCRIPTION
**Pull request by darken99**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cid/helpers/quicksight/__init__.py [hunk 1]
```diff
@@ -341,7 +352,7 @@ def list_data_sources(self) -> list:
                 return result.get('DataSources')
         except self.client.exceptions.AccessDeniedException:
             logger.info('Access denied listing data sources')
-            return list()
+            raise
         except Exception as e:
             logger.debug(e, stack_info=True)
             return list()
```
Comment (iakov-aws) [inline cid/helpers/quicksight/__init__.py:358]: Would it be more logical to raise here as well?

File: cid/common.py [hunk 2]
```diff
@@ -924,7 +920,7 @@ def create_or_update_dataset(self, dataset_definition: dict, arn: str=None,recur
 
         # create missing views
         if len(missing_views):
-            print(f'\tMissing Athena views: {missing_views}')
+            print(f"Missing views: {', '.joinmissing_views}")
```
Comment (iakov-aws) [inline cid/common.py:923]: joinmissing_views undefined

**Review by iakov-aws — CHANGES_REQUESTED** [review]

File: cid/helpers/quicksight/__init__.py (hunk 1 above)
Comment (darken99) [reply cid/helpers/quicksight/__init__.py:358]: Maybe, but will requires to change in other places as well

File: cid/common.py [hunk 3]
```diff
@@ -896,14 +890,16 @@ def create_or_update_dataset(self, dataset_definition: dict, arn: str=None,recur
         # else user will be suggested to choose database
 
         columns_tpl = {
-            'athena_datasource_arn': next(iter(self.qs.athena_datasources)),
+            'athena_datasource_arn': next(iter(v.arn for v in self.qs.athena_datasources.values())),
```
Comment (iakov-aws) [inline cid/common.py:893]: can we add a fallback here? 
```
'athena_datasource_arn': next(iter(v.arn for v in self.qs.athena_datasources.values()), None),
```

**Review by iakov-aws — APPROVED** [review]

File: cid/helpers/quicksight/__init__.py [hunk 4]
```diff
@@ -572,11 +583,14 @@ def describe_dataset(self, id, timeout: int=1) -> dict:
 
         return self._datasets.get(id, None)
 
-    def discover_datasets(self):
+    def discover_datasets(self, _datasets: list=None):
```
Comment (iakov-aws) [inline cid/helpers/quicksight/__init__.py:586]: suggest following

```python
  def discover_datasets(self, datasets: list=None):
        """ Discover datasets in the account 
        :datasets : list of datast ids, if not provided or empty, will discover all
        """

        logger.info('Discovering datasets')
        self._datasets =  self._datasets or {}

        try:
            dataset_ids = datasets or [d.get('DataSetId') for d in self.list_data_sets()]:
            for dataset in dataset_ids:
                try:
                    self.describe_dataset(dataset.get('DataSetId'))
                except Exception as e:
                    logger.debug(e, stack_info=True)
                    continue
        except self.client.exceptions.AccessDeniedException:
            logger.info('Access denied listing datasets')
        except Exception as e:
            logger.debug(e, stack_info=True)
            logger.info('No datasets found')
```

**Pull request merged**